In [1]:
import tensorflow as tf

In [2]:
mnist=tf.keras.datasets.mnist

In [3]:
from sklearn.model_selection import KFold

In [6]:
kf=KFold(n_splits=5)

In [7]:
kf.get_n_splits()

5

In [17]:
(x_train, y_train),(x_test,y_test)=mnist.load_data()

In [18]:
x_train,x_test=x_train/255.0,x_test/255.0

In [21]:
for train_index, val_index in kf.split(x_train):
    print("TRAIN:", train_index, "VALIDATE:", val_index)

TRAIN: [12000 12001 12002 ... 59997 59998 59999] VALIDATE: [    0     1     2 ... 11997 11998 11999]
TRAIN: [    0     1     2 ... 59997 59998 59999] VALIDATE: [12000 12001 12002 ... 23997 23998 23999]
TRAIN: [    0     1     2 ... 59997 59998 59999] VALIDATE: [24000 24001 24002 ... 35997 35998 35999]
TRAIN: [    0     1     2 ... 59997 59998 59999] VALIDATE: [36000 36001 36002 ... 47997 47998 47999]
TRAIN: [    0     1     2 ... 47997 47998 47999] VALIDATE: [48000 48001 48002 ... 59997 59998 59999]


In [23]:
model= tf.keras.models.Sequential([
    tf.keras.layers.Flatten(input_shape=(28,28)),
    tf.keras.layers.Dense(128, activation='relu'),
    tf.keras.layers.Dropout(0.2),
    tf.keras.layers.Dense(10)
])

In [25]:
for train_index, val_index in kf.split(x_train):
    print("train: ", train_index, "val: ", val_index)

train:  [12000 12001 12002 ... 59997 59998 59999] val:  [    0     1     2 ... 11997 11998 11999]
train:  [    0     1     2 ... 59997 59998 59999] val:  [12000 12001 12002 ... 23997 23998 23999]
train:  [    0     1     2 ... 59997 59998 59999] val:  [24000 24001 24002 ... 35997 35998 35999]
train:  [    0     1     2 ... 59997 59998 59999] val:  [36000 36001 36002 ... 47997 47998 47999]
train:  [    0     1     2 ... 47997 47998 47999] val:  [48000 48001 48002 ... 59997 59998 59999]


In [26]:
kf.split(x_train)

<generator object _BaseKFold.split at 0x000001CC5E7AC4C8>

In [38]:
predictions=model(x_train[:5]).numpy()
predictions

array([[-12.697265  ,  -6.4635673 ,  -5.5225396 ,   9.730963  ,
        -29.257755  ,  12.041705  , -16.24395   ,  -8.398741  ,
        -12.192362  ,  -4.441319  ],
       [ 14.414392  ,  -9.958658  ,   1.0502868 ,  -4.2826614 ,
        -13.989229  ,  -3.37674   ,  -3.2270741 ,  -2.770842  ,
         -6.96792   ,  -3.9350698 ],
       [ -7.178491  ,  -7.6510854 ,  -0.14363433,  -1.5341163 ,
          6.7151766 ,  -6.2901297 ,  -3.1360998 ,  -1.1705166 ,
         -2.918718  ,   1.6232035 ],
       [ -8.735497  ,  10.310317  ,  -3.8960068 ,  -7.1638236 ,
         -1.9973284 ,  -8.665093  ,  -7.6515727 ,  -0.18701401,
         -0.45610297, -11.002445  ],
       [-10.638469  ,  -5.129674  ,  -6.795559  ,   0.35584736,
         -0.03325601,  -5.878095  , -12.770574  ,  -1.291327  ,
         -0.7552044 ,   6.6925783 ]], dtype=float32)

In [40]:
print(y_train[:5])

[5 0 4 1 9]


In [39]:
tf.nn.softmax(predictions).numpy()

array([[1.6403255e-11, 8.3596996e-09, 2.1422657e-08, 9.0237170e-02,
        1.0539013e-18, 9.0976268e-01, 4.7274158e-13, 1.2071345e-09,
        2.7177340e-11, 6.3159909e-08],
       [9.9999845e-01, 2.5996682e-11, 1.5705153e-06, 7.5852871e-09,
        4.6180984e-13, 1.8767608e-08, 2.1797563e-08, 3.4399093e-08,
        5.1734339e-10, 1.0738138e-08],
       [9.1752020e-07, 5.7196593e-07, 1.0418734e-03, 2.5938000e-04,
        9.9210733e-01, 2.2306217e-06, 5.2264124e-05, 3.7311754e-04,
        6.4954802e-05, 6.0973726e-03],
       [5.3516094e-09, 9.9994600e-01, 6.7647005e-07, 2.5766409e-08,
        4.5168317e-06, 5.7419642e-09, 1.5820753e-08, 2.7608559e-05,
        2.1095004e-05, 5.5457600e-10],
       [2.9616183e-08, 7.3108567e-06, 1.3819224e-06, 1.7631933e-03,
        1.1948529e-03, 3.4588625e-06, 3.5121070e-09, 3.3957927e-04,
        5.8046542e-04, 9.9610978e-01]], dtype=float32)

In [30]:
loss_fn=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)

In [31]:
loss_fn(y_train[:1],predictions).numpy()

2.2318947

In [35]:
model.compile(optimizer='adam',loss=loss_fn,metrics=['accuracy'])

In [36]:
model.fit(x_train,y_train,epochs=5)

Epoch 1/5
1875/1875 [==============================] - 3s 1ms/step - loss: 0.2973 - accuracy: 0.9138
Epoch 2/5
1875/1875 [==============================] - 2s 1ms/step - loss: 0.1456 - accuracy: 0.9570
Epoch 3/5
1875/1875 [==============================] - 2s 1ms/step - loss: 0.1072 - accuracy: 0.9669
Epoch 4/5
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0880 - accuracy: 0.9728
Epoch 5/5
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0755 - accuracy: 0.9768


In [37]:
model.evaluate(x_test,y_test,verbose=2)

313/313 - 0s - loss: 0.0719 - accuracy: 0.9787


[0.07187596708536148, 0.9786999821662903]

The end accuracy of this test was 0.9787 compared to 0.9496 when not using KFold. 

The predictions for the first 5:
array([[1.6403255e-11, 8.3596996e-09, 2.1422657e-08, 9.0237170e-02,
        1.0539013e-18, 9.0976268e-01, 4.7274158e-13, 1.2071345e-09,
        2.7177340e-11, 6.3159909e-08],
       [9.9999845e-01, 2.5996682e-11, 1.5705153e-06, 7.5852871e-09,
        4.6180984e-13, 1.8767608e-08, 2.1797563e-08, 3.4399093e-08,
        5.1734339e-10, 1.0738138e-08],
       [9.1752020e-07, 5.7196593e-07, 1.0418734e-03, 2.5938000e-04,
        9.9210733e-01, 2.2306217e-06, 5.2264124e-05, 3.7311754e-04,
        6.4954802e-05, 6.0973726e-03],
       [5.3516094e-09, 9.9994600e-01, 6.7647005e-07, 2.5766409e-08,
        4.5168317e-06, 5.7419642e-09, 1.5820753e-08, 2.7608559e-05,
        2.1095004e-05, 5.5457600e-10],
       [2.9616183e-08, 7.3108567e-06, 1.3819224e-06, 1.7631933e-03,
        1.1948529e-03, 3.4588625e-06, 3.5121070e-09, 3.3957927e-04,
        5.8046542e-04, 9.9610978e-01]], dtype=float32)